Revise zooming to get dialogs to position correctly (BL-4172)

StephenMcConnel opened the pull request:
This involved moving the transform:scale css from the body element to a
outer div element created especially for holding the scale factor.  With
this fix, the dialogs and dropdowns just positioned themselves nicely,
and the tooltips needed only a minor tweak to continue working as before.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1610)
<!-- Reviewable:end -->

JohnThomson commented:
This is probably inherited from a mistake I made, but while we're in this area how about cleaning up by consistently changing scale(x,x) to scale(x) which does the same?
Otherwise, just a few minor suggestions for improving names and comments and so forth

---

Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/js/BloomHintBubbles.ts, line 238 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgEyfCo-QACQWo16Xkw:-KgEyfCo-QACQWo16Xkx:b-jro5vs) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomBrowserUI/bookEdit/js/BloomHintBubbles.ts#L238)):*
> ```XML
>             viewport: $(window),
>             adjust: { method: 'none' },
>             container: $("div#page-scaling-container")
> ```

I think this deserves a comment...there are specific reasons for putting it in this div rather than the body or the page element.
I wonder whether we could tie all the qtips that need to go here together with a method somewhere like qtipContainer()?

---

*[src/BloomExe/Edit/EditingModel.cs, line 648 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgEzkm3hcvl7ZqDvrNi:-KgEzkm3hcvl7ZqDvrNj:blab6dh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L648)):*
> ```Smalltalk
> 		/// Set a style on the body of the main content page that will zoom it to the extent the user currently prefers.
> 		/// </summary>
> 		private void SetPageZoom()
> ```

This method is now also responsible for inserting the wrap div. Consider make the caller (probably only one?) separately call InsertContainingScalingDiv, or improving the name of this...maybe SetupZoom()?

---

*[src/BloomExe/Edit/EditingModel.cs, line 661 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF0En42_nEkvO0rAsk:-KgF0En42_nEkvO0rAsl:b-et3qel) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L661)):*
> ```Smalltalk
> 		}
> 
> 		XmlElement InsertContainingScalingDiv(XmlElement body, XmlElement pageDiv)
> ```

This is a little messy and non-obvious, so it would be good to capture in a comment why we did it...problems with dialogs and pull-downs if we put scale on body, and with qtips if we put it on the page. This feels like a good place to put such a comment.

---

*[src/BloomExe/Edit/EditingModel.cs, line 939 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF0A32d3ThGYd6ocfm:-KgF0A33Xh8XfZ4Ru4Fb:by1cae7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L939)):*
> ```Smalltalk
> 		/// Currently this is just the zoom level.
> 		/// </summary>
> 		void SavePageFrameState()
> ```

Again, this method is now doing more than just save the frame state.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-:-KgF0w3okhwry68AeNcP:b-m0a10c)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomExe/Edit/EditingModel.cs, line 976 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF3pSn8A1KP6JvZqTz:-KgF3pSn8A1KP6JvZqU-:bl7ptq1) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L976)):*
> testForPageScalingDiv

GetPageScalingDiv?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
done.

---

Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/BloomBrowserUI/bookEdit/js/BloomHintBubbles.ts, line 238 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgEyfCo-QACQWo16Xkw:-KgFQ-pu3oOvayzAx8jT:b-tdtoo3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomBrowserUI/bookEdit/js/BloomHintBubbles.ts#L238)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I think this deserves a comment...there are specific reasons for putting it in this div rather than the body or the page element.
I wonder whether we could tie all the qtips that need to go here together with a method somewhere like qtipContainer()?
</blockquote></details>

Done.  I actually named it qtipZoomContainer

---

*[src/BloomExe/Edit/EditingModel.cs, line 648 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgEzkm3hcvl7ZqDvrNi:-KgFTij06HSA80elvfXk:b-vg5oh0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L648)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This method is now also responsible for inserting the wrap div. Consider make the caller (probably only one?) separately call InsertContainingScalingDiv, or improving the name of this...maybe SetupZoom()?
</blockquote></details>

Renamed to  SetupPageZoom and changed the method summary comment to explain what is happening and why.

---

*[src/BloomExe/Edit/EditingModel.cs, line 661 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF0En42_nEkvO0rAsk:-KgFTwG88-FX2dGdypgK:b8pq1j1) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L661)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This is a little messy and non-obvious, so it would be good to capture in a comment why we did it...problems with dialogs and pull-downs if we put scale on body, and with qtips if we put it on the page. This feels like a good place to put such a comment.
</blockquote></details>

I tried to explain the in the method's summary comment.

---

*[src/BloomExe/Edit/EditingModel.cs, line 939 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF0A32d3ThGYd6ocfm:-KgFb-Dl3P4F-pARGUiW:b5qhgnl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L939)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Again, this method is now doing more than just save the frame state.
</blockquote></details>

I renamed the method and added some comments.

---

*[src/BloomExe/Edit/EditingModel.cs, line 976 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF3pSn8A1KP6JvZqTz:-KgFR6u2nAn1aYfMpj7b:b-17df9g) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L976)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> testForPageScalingDiv

GetPageScalingDiv?
</blockquote></details>

It doesn't really get the div, it's a Predicate<GeckoElement> for testing whether it's a desired div.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-:-KgFMLq8QOIaJVnkdDuk:b-7rl0jd)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/BloomExe/Edit/EditingModel.cs, line 976 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF3pSn8A1KP6JvZqTz:-KgFuMUkI35djgjV_o_1:b-4xzjlt) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L976)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

It doesn't really get the div, it's a Predicate<GeckoElement> for testing whether it's a desired div.
</blockquote></details>

Yep. Totally missed that. I see now it is a predicate. The method name should still be capitalized, right?

Also, it is probably worth making page-scaling-container a constant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Much better...couple of minor points more to think about

---

Review status: 0 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Edit/EditingModel.cs, line 976 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF3pSn8A1KP6JvZqTz:-KgGFj2oeuziC_-wXb02:b-jxwk3a) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L976)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Yep. Totally missed that. I see now it is a predicate. The method name should still be capitalized, right?

Also, it is probably worth making page-scaling-container a constant.
</blockquote></details>

Maybe IsPageScalingDiv()? I think that's a more common pattern for "Is this an X?" test... made me think this was something to do with unit tests. (Speaking of which, is there any of this we could reasonably unit-test?)

---

*[src/BloomExe/Edit/EditingModel.cs, line 945 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgGEDB-lAE5JEYj9egh:-KgGEDB0F7_qVB1Q27Zm:bg1r8mo) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1ed6122b076a9bb8c74b44caf0b726027e8226be/src/BloomExe/Edit/EditingModel.cs#L945)):*
> ```Smalltalk
> 		/// Save anything we want to persist from page to page but which is not part of the book from the page's current state.
> 		/// Currently this is just the zoom level.
> 		/// Also cleanup any html inserted for the page that we don't want to save.  Currently, this is just the div inserted
> ```

This isn't quite accurate. HtmlDom.ProcessPageAfterEditing() also has an important role in removing stuff we don't want to save. Would it be accurate to say that this removes stuff added by EditingModel code that we don't want to save? And maybe cross-ref HtmlDom.ProcessPageAfterEditing()? May be worth adding a cross-ref there to this method, too.
Sorry to be picky, but I suspect this could easily become a source of confusion, either of the "what on earth is removing the bloom-ui elements?" kind or "where should I add this new thing that needs cleaning up?" kind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-:-KgGG8nn_TGM0pDuK3gU:bo90urw)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/Edit/EditingModel.cs, line 976 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgF3pSn8A1KP6JvZqTz:-KgKXSlL99lOqp5i6vPQ:bdjhk5l) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/2d8cbd80829a6555a11c9c5d96cc069c5490d1b1/src/BloomExe/Edit/EditingModel.cs#L976)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Maybe IsPageScalingDiv()? I think that's a more common pattern for "Is this an X?" test... made me think this was something to do with unit tests. (Speaking of which, is there any of this we could reasonably unit-test?)
</blockquote></details>

Done.  Renamed to IsPageScalingDiv.  "page-scaling-container" is now a constant in C# code.

---

*[src/BloomExe/Edit/EditingModel.cs, line 945 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-KgGEDB-lAE5JEYj9egh:-KgKfTSSOPXk0LRdiy9k:b-kq2xup) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/1ed6122b076a9bb8c74b44caf0b726027e8226be/src/BloomExe/Edit/EditingModel.cs#L945)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This isn't quite accurate. HtmlDom.ProcessPageAfterEditing() also has an important role in removing stuff we don't want to save. Would it be accurate to say that this removes stuff added by EditingModel code that we don't want to save? And maybe cross-ref HtmlDom.ProcessPageAfterEditing()? May be worth adding a cross-ref there to this method, too.
Sorry to be picky, but I suspect this could easily become a source of confusion, either of the "what on earth is removing the bloom-ui elements?" kind or "where should I add this new thing that needs cleaning up?" kind.
</blockquote></details>

Done.  I've revised the comment and added link comments elsewhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610#-:-KgL88_39wAFOFBEv-RJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Reviewed 8 of 9 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1610)*
<!-- Sent from Reviewable.io -->
